storage/innobase/log/log0recv.cc

```diff
@@ -1640,11 +1639,11 @@ static byte *recv_parse_or_apply_log_rec_body(mlog_id_t type, byte *ptr,
                 fil_tablespace_redo_encryption(ptr, end_ptr, space_id, apply));
           } else if (memcmp(ptr_copy, ENCRYPTION_KEY_MAGIC_PS_V1,
                             ENCRYPTION_MAGIC_SIZE) == 0 &&
-                     apply) {
+                     !apply) {
             return (fil_parse_write_crypt_data_v1(space_id, ptr, end_ptr, len));
           } else if (memcmp(ptr_copy, ENCRYPTION_KEY_MAGIC_PS_V2,
                             ENCRYPTION_MAGIC_SIZE) == 0 &&
-                     apply) {
+                     !apply) {
             return (fil_parse_write_crypt_data_v2(space_id, ptr, end_ptr, len));
```

Comment:
So, the apply is used to tell whether to apply a record to the page or not. and it should be false for the 'parse' phase which you have done below.. 

For MK, after a long study, I found out that this fil_tablespace_encryption_redo() is called three times :open_mouth: 
2 times during parse and 1 time from the actual apply.

1st iteration (parse): space is nullptr, so it stores in recv_sys -> keys
2 iteration (parse): space is  nullptr again, wasted work, updates the same key again for key which is already stored in recv_sys keys
3rd iteration (apply from recover_page_func()):  sets space->encryption_key and iv from recv_sys->keys.

There is also encryption decoding when we parse the encryption_data from redo.

WIth your fix, you seem to parse (ie. do fil_parse_crypt_data_v2) during the parse stage and write to page 0 on apply.. 

This looks correct but I think there is issue here
Would there be space->crypt_data during the apply stage? During apply stage, when the space is **not** nullptr, we do not call fil_parse(). So space->crypt_data should be null ptr. This would be necessary for decryption of other pages. isn't it?

Reply:
It is not called during apply stage, but is called during parse stage. So when tablespace already exists in the cache - it will get crypt_data assigned. In case it does not exist (create table is only in redo) only changes to page0 will get applied (crypt_data will not be assigned in log0recv). And all other changes to other pages will get applied (redo is not encrypted, and all pages associated with the table are still in redo). After apply the table will be opened by open_for_recovery (IIRC), space will be loaded into cache and crypt_data will get read from page0.

Reply:
I understood the fix, During parse (apply is false), we just parse the crypt_data in the redo. (if space exists, crypt_data gets assigned), During apply (apply is true), we make the modification to page0.

In the scenario where page 0 changes are only in redo, and changes on other pages need decryption or encryption.
In this case, since space->crypt_data is nullptr, all the pages written during the redo stage will be written either MK encrypted or unencrypted.

Consider calling the parse + page 0 write during the apply (apply is true).

Reply:
I have made a mistake when I was trying to understand how this work. Please check the new commit message. I have changed the fix. Now crypt_data is assigned during opening tablespace for recovery. All changes from redo will follow the encryption in crypt_data.